File: packages/train-lle.js/src/core/engine.ts
import { Tensor } from "./tensor";
import { Optimizer } from "./optimizer";
import { Loss } from "./loss";

export interface Layer {
  type: string;
  config: any;
  forward(input: Tensor): Tensor;
  backward(grad: Tensor): Tensor;
  params(): Tensor[];
  grads(): Tensor[];
}

export class Model {
  layers: Layer[];

  constructor(layers: Layer[]) {
    this.layers = layers;
  }

  forward(input: Tensor): Tensor {
    let out = input;
    for (const layer of this.layers) {
      out = layer.forward(out);
    }
    return out;
  }

  backward(grad: Tensor): void {
    let g = grad;
    for (let i = this.layers.length - 1; i >= 0; i--) {
      g = this.layers[i].backward(g);
    }
  }

  params(): Tensor[] {
    return this.layers.flatMap(l => l.params());
  }

  grads(): Tensor[] {
    return this.layers.flatMap(l => l.grads());
  }

  predict(input: Tensor): Tensor {
    return this.forward(input);
  }
}

export class Trainer {
  optimizer: Optimizer;
  lossFn: Loss;
  epochs: number;

  constructor(config: { optimizer: Optimizer; loss: Loss; epochs: number }) {
    this.optimizer = config.optimizer;
    this.lossFn = config.loss;
    this.epochs = config.epochs;
  }

  fit(model: Model, inputs: Tensor, targets: Tensor): void {
    for (let epoch = 0; epoch < this.epochs; epoch++) {
      const pred = model.forward(inputs);
      const loss = this.lossFn.forward(pred, targets);
      const grad = this.lossFn.backward(pred, targets);
      model.backward(grad);
      this.optimizer.step(model.params(), model.grads());
      console.log(`Epoch ${epoch + 1}/${this.epochs}, Loss: ${loss}`);
    }
  }

  evaluate(model: Model, inputs: Tensor, targets: Tensor): number {
    const pred = model.forward(inputs);
    return this.lossFn.forward(pred, targets);
  }
}

File: archive/train-lle-py/src/train_lle/engine.py
from .tensor import Tensor
import numpy as np

class Layer:
    def forward(self, input: Tensor) -> Tensor:
        raise NotImplementedError

    def backward(self, grad: Tensor) -> Tensor:
        raise NotImplementedError

    def params(self):
        return []

    def grads(self):
        return []

class Model:
    def __init__(self, layers):
        self.layers = layers

    def forward(self, input: Tensor) -> Tensor:
        out = input
        for layer in self.layers:
            out = layer.forward(out)
        return out

    def backward(self, grad: Tensor):
        g = grad
        for layer in reversed(self.layers):
            g = layer.backward(g)

    def params(self):
        return [p for layer in self.layers for p in layer.params()]

    def grads(self):
        return [g for layer in self.layers for g in layer.grads()]

    def predict(self, input: Tensor) -> Tensor:
        return self.forward(input)

class Trainer:
    def __init__(self, optimizer, loss, epochs):
        self.optimizer = optimizer
        self.loss_fn = loss
        self.epochs = epochs

    def fit(self, model: Model, inputs: Tensor, targets: Tensor):
        for epoch in range(self.epochs):
            pred = model.forward(inputs)
            loss = self.loss_fn.forward(pred, targets)
            grad = self.loss_fn.backward(pred, targets)
            model.backward(grad)
            self.optimizer.step(model.params(), model.grads())
            print(f"Epoch {epoch + 1}/{self.epochs}, Loss: {loss}")

    def evaluate(self, model: Model, inputs: Tensor, targets: Tensor):
        pred = model.forward(inputs)
        return self.loss_fn.forward(pred, targets)